pkg/etcd-backup-operator/controller/s3backup.go

```diff
@@ -0,0 +1,233 @@
+package controller
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+
+	api "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+	"github.com/coreos/etcd-operator/pkg/util/constants"
+	"github.com/coreos/etcd-operator/pkg/util/etcdutil"
+	"github.com/coreos/etcd-operator/pkg/util/k8sutil"
+	"github.com/coreos/etcd/clientv3"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	S3V1                 = "v1"
+	operatorRoot         = "/var/tmp/etcd-backup-operator"
+	backupFilenameSuffix = "etcd.backup"
+)
+
+func (b *Backup) handleS3(clusterName string, s3 *api.S3Source) error {
+	s3b := &s3Backup{
+		namespace:   b.namespace,
+		name:        b.name,
+		clusterName: clusterName,
+		s3bucket:    s3.S3Bucket,
+		prefix:      ToS3Prefix(s3.Prefix, b.namespace, clusterName),
+		awsSecret:   s3.AWSSecret,
+		kubecli:     b.kubecli,
+	}
+	return s3b.saveSnap()
+}
+
+// ToS3Prefix concatenates s3Prefix, S3V1, namespace, clusterName to a single s3 prefix.
+// the concatenated prefix determines the location of S3 backup files.
+func ToS3Prefix(s3Prefix, namespace, clusterName string) string {
+	return path.Join(s3Prefix, S3V1, namespace, clusterName)
+}
+
+type s3Backup struct {
+	namespace   string
+	name        string
+	clusterName string
+
+	s3bucket  string
+	prefix    string
+	awsSecret string
+
+	kubecli kubernetes.Interface
+}
+
+func (b *s3Backup) saveSnap() error {
```

Comment:
We're already using the receiver `(b *Backup)` above.
Plus the naming convention for method receivers is that generally it should be the first one or two letters of its type. So can you rename them to `(s *s3Backup)`

Reply:
agreed.